### PR TITLE
Issue #39: create function to write iostats to file

### LINKED ISSUE #39: Collect status of io devices pre-termination
This should be a [node-level](https://github.com/tetratelabs/getenvoy/blob/master/pkg/binary/envoy/debug/node.go) collection. Output should look similar to (or the same as) iostat.
The aim is to keep external dependencies to a minimum, so [this library](https://godoc.org/github.com/shirou/gopsutil/disk) should be used.

```sh
iostat -dxm
Linux 4.9.184-linuxkit (4f51979ee734) 	08/14/19 	_x86_64_	(4 CPU)

Device            r/s     w/s     rMB/s     wMB/s   rrqm/s   wrqm/s  %rrqm  %wrqm r_await w_await aqu-sz rareq-sz wareq-sz  svctm  %util
scd0             0.01    0.00      0.00      0.00     0.00     0.00   0.00   0.00    0.72    0.00   0.00    73.45     0.00   0.63   0.00
scd1             0.00    0.00      0.00      0.00     0.00     0.00   0.00   0.00    0.00    0.00   0.00    17.71     0.00   0.00   0.00
scd2             0.01    0.00      0.00      0.00     0.00     0.00   0.00   0.00    0.51    0.00   0.00    70.44     0.00   0.42   0.00
sda              0.23    0.26      0.01      0.01     0.00     0.18   0.35  40.50    0.40    3.09   0.00    31.96    56.46   0.25   0.01
```

### PR DESCRIPTION
**Pull request by jhengy**
# Task accomplished
- fixes #39 
- function invoked in pre-termination: write iostat from [this function](https://godoc.org/github.com/shirou/gopsutil/disk#IOCountersStat) to a json file
- example file written:
```json
[
 ...
  "loop9": {
    "readCount": 20,
    "mergedReadCount": 0,
    "writeCount": 0,
    "mergedWriteCount": 0,
    "readBytes": 48128,
    "writeBytes": 0,
    "readTime": 0,
    "writeTime": 0,
    "iopsInProgress": 0,
    "ioTime": 0,
    "weightedIO": 0,
    "name": "loop9",
    "serialNumber": "",
    "label": ""
  },
  "nvme0n1p1": {
    "readCount": 1089,
    "mergedReadCount": 0,
    "writeCount": 2,
    "mergedWriteCount": 0,
    "readBytes": 6055424,
    "writeBytes": 1024,
    "readTime": 1032,
    "writeTime": 0,
    "iopsInProgress": 0,
    "ioTime": 24,
    "weightedIO": 976,
    "name": "nvme0n1p1",
    "serialNumber": "",
    "label": ""
  },
  "nvme0n1p5": {
    "readCount": 482911,
    "mergedReadCount": 43268,
    "writeCount": 179516,
    "mergedWriteCount": 488437,
    "readBytes": 6519100416,
    "writeBytes": 5955575808,
    "readTime": 143596,
    "writeTime": 183952,
    "iopsInProgress": 0,
    "ioTime": 23972,
    "weightedIO": 195900,
    "name": "nvme0n1p5",
    "serialNumber": "",
    "label": ""
  }
]
```

Signed-off-by: Hengyuan <jhy8890948@gmail.com>

### CODE REVIEW

File: pkg/binary/envoy/debug/node.go [hunk 1]
```diff
@@ -22,6 +22,7 @@ import (
 	"path/filepath"
 	"text/tabwriter"
 
+	d "github.com/shirou/gopsutil/disk"
```
Comment (liamawhite) [inline pkg/binary/envoy/debug/node.go]: Actually can you also just remove the name for this package and use `disk`.

File: pkg/binary/envoy/debug/node.go [hunk 2]
```diff
@@ -141,3 +143,29 @@ func networkInterfaces(r binary.Runner) error {
...
+func writeIOStats(r binary.Runner) error {
+	f, err := os.Create(filepath.Join(r.DebugStore(), "node/iostats.json"))
+	defer f.Close() //nolint
+	if err != nil {
+		return fmt.Errorf("error in creating iostat.json: %v", err)
+	}
+
+	physicalPartitions, _ := d.Partitions(false)
+	deviceNames := make([]string, 0, len(physicalPartitions))
+	//nolint:gocritic
```
Comment (liamawhite) [inline pkg/binary/envoy/debug/node.go]: What was the complaint here? CopyRange?

Comment (jhengy) [reply pkg/binary/envoy/debug/node.go]: Yes, the error `rangeValCopy: each iteration copies 64 bytes (consider pointers or indexing) (gocritic)` is returned

File: pkg/binary/envoy/debug/node.go [hunk 3]
```diff
@@ -141,3 +143,29 @@ func networkInterfaces(r binary.Runner) error {
...
+
+// writeIOStat write iostat of devices in the form of a dictionary to json file
+func writeIOStats(r binary.Runner) error {
+	f, err := os.Create(filepath.Join(r.DebugStore(), "node/iostats.json"))
+	defer f.Close() //nolint
+	if err != nil {
+		return fmt.Errorf("error in creating iostat.json: %v", err)
+	}
+
+	physicalPartitions, _ := d.Partitions(false)
```
Comment (liamawhite) [inline pkg/binary/envoy/debug/node.go]: Is this an error? We should handle it.

File: pkg/binary/envoy/debug/node.go (hunk 2 above)
Comment (taiki45) [reply pkg/binary/envoy/debug/node.go]: nit: Should we ignore the lint error? I'm not familiar with Golang, I think we should follow linters in most of the cases, here we can avoid the copy like:

```
for i := range physicalPartitions {
  deviceNames = append(deviceNames, physicalPartitions[i].Device)
}
```

Comment (jhengy) [reply pkg/binary/envoy/debug/node.go]: Thank you. That's a really good suggestion.

File: pkg/binary/envoy/debug/node_test.go [hunk 4]
```diff
@@ -34,7 +34,8 @@ func Test_debugging_outputs(t *testing.T) {
 		defer os.RemoveAll(r.DebugStore())
 		envoytest.RunKill(r, filepath.Join("testdata", "null.yaml"), time.Second*10)
 
-		files := [...]string{"node/ps.txt", "node/network_interface.json", "node/connections.json"}
+		files := [...]string{"node/ps.txt", "node/network_interface.json", "node/connections.json", "node/iostats.json"}
```
Comment (ZackButcher) [inline pkg/binary/envoy/debug/node_test.go]:
```suggestion
		files := []string{"node/ps.txt", "node/network_interface.json", "node/connections.json", "node/iostats.json"}
```

I don't know if I've seen this `[...]` syntax before; either way it's not needed, let's just remove

**Comment by mathetake** [comment]
closing since it's stabled.

**Pull request closed without merging**